File: superstitioMain/java/superstitio/customStrings/stringsSet/OrbStringsSet.java
package superstitio.customStrings.stringsSet;

import com.megacrit.cardcrawl.localization.LocalizedStrings;
import com.megacrit.cardcrawl.localization.OrbStrings;
import superstitio.customStrings.interFace.HasOriginAndSFWVersion;
import superstitio.customStrings.interFace.StringSetUtility;
import superstitio.customStrings.interFace.WordReplace;

import java.util.List;

public class OrbStringsSet implements HasOriginAndSFWVersion<OrbStrings> {
    private final OrbStrings Origin = new OrbStrings();
    private final OrbStrings SFW = new OrbStrings();
    private String NAME;
    private String NAME_SFW;
    private String[] DESCRIPTION;
    private String[] DESCRIPTION_SFW;

    public OrbStringsSet() {
    }

    public String getNAME() {
        return getFromRightVersion(strings -> strings.NAME);
    }

    public String[] getDESCRIPTION() {
        return getArrayFromRightVersion(strings -> strings.DESCRIPTION);
    }

    @Override
    public void initialSelfBlack() {
        this.NAME = "[MISSING_NAME]";
        this.DESCRIPTION = LocalizedStrings.createMockStringArray(10);
    }

    @Override
    public void initialOrigin(OrbStrings origin) {
        origin.NAME = NAME;
        origin.DESCRIPTION = DESCRIPTION;
    }

    @Override
    public void initialSFW(OrbStrings sfw) {
        sfw.NAME = NAME_SFW;
        sfw.DESCRIPTION = DESCRIPTION_SFW;
    }

    public OrbStrings getRightVersion() {
        if (StringSetUtility.shouldReturnSFWVersion(SFW.NAME))
            return SFW;
        return Origin;
    }

    @Override
    public Class<OrbStrings> getSubClass() {
        return OrbStrings.class;
    }

    @Override
    public OrbStrings getSFWVersion() {
        return SFW;
    }

    @Override
    public OrbStrings getOriginVersion() {
        return Origin;
    }

    @Override
    public void setupSFWStringByWordReplace(List<WordReplace> replaceRules) {
//        replaceWord_NAME(this.SFW, replaceRules);
//        replaceWord_DESCRIPTION(this.SFW, replaceRules);

        if (StringSetUtility.isNullOrEmpty(this.SFW.NAME))
            this.SFW.NAME = WordReplace.replaceWord(this.Origin.NAME, replaceRules);
        if (StringSetUtility.isNullOrEmpty(this.SFW.DESCRIPTION))
            this.SFW.DESCRIPTION = WordReplace.replaceWord(this.Origin.DESCRIPTION, replaceRules);
    }

//    private void replaceWord_NAME(OrbStrings replaced, List<WordReplace> replaceRules) {
//        if (StringSetUtility.isNullOrEmpty(replaced.NAME))
//            replaced.NAME = WordReplace.replaceWord(this.Origin.NAME, replaceRules);
//        else
//            replaced.NAME = WordReplace.replaceWord(replaced.NAME, replaceRules);
//    }

//    private void replaceWord_DESCRIPTION(OrbStrings replaced, List<WordReplace> replaceRules) {
//        if (StringSetUtility.isNullOrEmpty(replaced.DESCRIPTION))
//            replaced.DESCRIPTION = WordReplace.replaceWord(this.Origin.DESCRIPTION, replaceRules);
//        else
//            WordReplace.replaceWord(replaced.DESCRIPTION, replaceRules);
//    }

//    public WordReplace toCardNameReplaceRule(){
//        return new WordReplace(this.NAME,this.NAME_SFW);
//    }
//
//    public static List<WordReplace> makeCardNameReplaceRules(List<PowerStringsWithSFW> cards){
//        return cards.stream().map(PowerStringsWithSFW::toCardNameReplaceRule).collect(Collectors.toList());
//    }
}

File: superstitioMain/java/superstitio/customStrings/stringsSet/OrbStringsSet.kt
package superstitio.customStrings.stringsSet

import com.megacrit.cardcrawl.localization.LocalizedStrings
import com.megacrit.cardcrawl.localization.OrbStrings
import superstitio.customStrings.interFace.*

class OrbStringsSet : HasOriginAndSFWVersion<OrbStrings>
{
    override val OriginVersion = OrbStrings()
    override val SfwVersion = OrbStrings()
    private var NAME: String? = null
    private var NAME_SFW: String? = null
    private var DESCRIPTION: Array<String>? = null
    private var DESCRIPTION_SFW: Array<String>? = null

    fun getNAME(): String
    {
        return getFromRightVersion { strings: OrbStrings? -> strings!!.NAME }!!
    }

    fun getDESCRIPTION(): Array<String>
    {
        return getArrayFromRightVersion { strings: OrbStrings? -> strings!!.DESCRIPTION }!!
    }

    override fun initialSelfBlack()
    {
        this.NAME = "[MISSING_NAME]"
        this.DESCRIPTION = LocalizedStrings.createMockStringArray(10)
    }

    override fun initialOrigin(origin: OrbStrings)
    {
        origin.NAME = NAME
        origin.DESCRIPTION = DESCRIPTION
    }

    override fun initialSFW(sfw: OrbStrings)
    {
        sfw.NAME = NAME_SFW
        sfw.DESCRIPTION = DESCRIPTION_SFW
    }

    override fun getRightVersion(): OrbStrings
    {
        if (StringSetUtility.shouldReturnSFWVersion(SfwVersion.NAME)) return SfwVersion
        return OriginVersion
    }

    override fun getSubClass() = OrbStrings::class.java

    override fun makeCopy(): HasDifferentVersionStringSet<OrbStrings>
    {
        val clone = OrbStringsSet()
        clone.NAME = this.NAME
        clone.DESCRIPTION = this.DESCRIPTION
        clone.NAME_SFW = this.NAME_SFW
        clone.DESCRIPTION_SFW = this.DESCRIPTION_SFW
        return clone
    }

    override fun setupSFWStringByWordReplace(replaceRules: List<WordReplace>)
    {
        SfwVersion.NAME =
            updateFieldIfEmpty(SfwVersion.NAME, OriginVersion.NAME, replaceRules)
        SfwVersion.DESCRIPTION =
            updateFieldIfEmpty(SfwVersion.DESCRIPTION, OriginVersion.DESCRIPTION, replaceRules)

        this.NAME_SFW = this.NAME_SFW.takeIfNullOrEmpty(SfwVersion.NAME)
        this.DESCRIPTION_SFW = this.DESCRIPTION_SFW.takeIfNullOrEmpty(SfwVersion.DESCRIPTION)
    }

    override fun makeSFWCopy(): HasDifferentVersionStringSet<OrbStrings>
    {
        val clone: HasDifferentVersionStringSet<OrbStrings> = this.makeCopy()
        if (clone is OrbStringsSet)
        {
            clone.NAME = null
            clone.DESCRIPTION = null
        }
        return clone
    }
}
